toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -132,6 +213,53 @@ func hasUnresolvedNodes(graph *pkggraph.PkgGraph) bool {
 	return false
 }
 
+func parseAndGeneratePackageList(dependencyGraph *pkggraph.PkgGraph, pkgsToBuild, pkgsToRebuild, pkgsToIgnore, imageConfig, baseDirPath string) (finalPackagesToBuild []*pkgjson.PackageVer, err error) {
+	// Generate the list of packages that need to be built.
+	// If none are requested then all packages will be built.
+	packagesToBuild, err := schedulerutils.PackageNamesToBuiltPackages(exe.ParseListArgument(pkgsToBuild), dependencyGraph)
+	if err != nil {
+		err = fmt.Errorf("unable to find build nodes for the packages to build, error:\n%w", err)
+		return
+	}
+
+	packagesToRebuild, err := schedulerutils.PackageNamesToBuiltPackages(exe.ParseListArgument(pkgsToRebuild), dependencyGraph)
+	if err != nil {
+		err = fmt.Errorf("unable to find build nodes for the packages to rebuild, error:\n%w", err)
+		return
+	}
+
+	prunedIgnoredPackageNames, unknownNames, err := schedulerutils.PruneUnknownPackages(exe.ParseListArgument(pkgsToIgnore), dependencyGraph)
+	if err != nil {
+		err = fmt.Errorf("failed to prune unknown package/spec names from the ignored list, error:\n%w", err)
+		return
+	}
+
+	if len(unknownNames) != 0 {
+		logger.Log.Warnf("The following ignored items matched neither a spec nor a package name: %v.", unknownNames)
+		return
+	}
+
+	packagesToIgnore, err := schedulerutils.PackageNamesToBuiltPackages(prunedIgnoredPackageNames, dependencyGraph)
+	if err != nil {
+		err = fmt.Errorf("unable to find build nodes for the ignored packages, error:\n%w", err)
+		return
+	}
+
+	ignoredAndRebuiltPackages := intersect.Hash(packagesToIgnore, packagesToRebuild)
+	if len(ignoredAndRebuiltPackages) != 0 {
+		err = fmt.Errorf("can't ignore and force a rebuild of a package at the same time. Abusing packages: %v", ignoredAndRebuiltPackages)
+		return
+	}
+
+	finalPackagesToBuild, err = schedulerutils.CalculatePackagesToBuild(packagesToBuild, packagesToRebuild, imageConfig, baseDirPath, dependencyGraph)
+	if err != nil {
+		err = fmt.Errorf("unable to generate package build list, error:\n%w", err)
+		return
+	}
+
+	return
```

Comment:
This looks like a lot of repeated code from the scheduler. I think we should refactor these steps into a separate tool, which can be used by both the package fetcher and the scheduler.

Reply:
I'll give it a look